src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java

```diff
@@ -707,6 +821,30 @@ private void onTokenExpiredSignal(
             }
         }
 
+        private void onTokenExpiringSoonSignal(
+            SignalFW signal)
+        {
+            final long finalDelay = grant.expiresAt - System.currentTimeMillis();
+            this.grantValidationFuture = executor.schedule(finalDelay, MILLISECONDS, targetRouteId, targetStreamId,
+                    GRANT_VALIDATION_SIGNAL);
```

Comment:
Move scheduling of the next signal back to `onTokenExpiredSignal` method, now all codepaths in the `if (delay > 0)` branch call `executor.schedule`, so combine them into one but use the conditional logic to determine the correct delay.